src/opossum_lib/tree_generation.py

```diff
@@ -0,0 +1,88 @@
+# SPDX-FileCopyrightText: 2023 TNG Technology Consulting GmbH <https://www.tngtech.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+from typing import Any, Dict, Optional, Tuple, Union
+
+from networkx import DiGraph, edge_bfs, is_weakly_connected, weakly_connected_components
+from spdx.model.document import CreationInfo
+from spdx.model.file import File
+from spdx.model.package import Package
+
+
+def generate_tree_from_graph(
+    graph: DiGraph,
+    source: Optional[str] = "SPDXRef-DOCUMENT",
+    tree: Optional[DiGraph] = None,
+) -> DiGraph:
+    if not graph.edges():
+        return graph
+
+    if is_weakly_connected(graph):
+        tree = tree or DiGraph()
+        edges_bfs = edge_bfs(graph, source)
+        visited_edges = []
+        for edge in edges_bfs:
+            _add_source_node(edge, graph, tree)
+            _add_target_node_and_edge(edge, graph, tree)
+            visited_edges += [edge]
+        # check if there is a directed path from source to each element node
+        # if not, we need to construct sub-graphs from the unreached edges
+        # and construct trees for all of these sub-graphs, the result
+        # will be an unconnected graph
+        unreached_edges = set(graph.edges).difference(set(visited_edges))
+        if unreached_edges:
+            induced_subgraph = graph.edge_subgraph(unreached_edges)
+            source = _get_node_without_incoming_edge(induced_subgraph)
+            tree_component = generate_tree_from_graph(induced_subgraph, source, tree)
+            tree.add_nodes_from(tree_component.nodes(data=True))
+            tree.add_edges_from(tree_component.edges(data=True))
+
+    else:  # get connected components
+        tree = DiGraph()
+        for connected_set in weakly_connected_components(
+            graph
+        ):  # returns only a set of nodes without edges
+            connected_subgraph = graph.subgraph(connected_set).copy()
+            # if the documents node is not in the subgraph we choose
+            # any elements node without incoming edge
+            source = (
+                "SPDXRef-DOCUMENT"
+                if "SPDXRef-DOCUMENT" in connected_set
+                else _get_node_without_incoming_edge(connected_subgraph)
+            )
+            tree_component = generate_tree_from_graph(connected_subgraph, source)
+            tree.add_nodes_from(tree_component.nodes(data=True))
+            tree.add_edges_from(tree_component.edges(data=True))
+
+    return tree
+
+
+def _add_target_node_and_edge(
+    edge: Tuple[str, str], graph: DiGraph, tree: DiGraph
+) -> None:
+    if edge[1] not in tree:
+        target_node_data = graph.nodes[edge[1]]
+        tree.add_node(edge[1], **target_node_data)
```

Comment:
These 3 lines could be extracted into a function `copy_node(node, source: DiGraph, target: DiGraph)` that could also be used in the `_add_source_node` function.

Reply:
I decided not to extract a method for these lines. I did extract one more method in the top-level function `_generate_tree_from_graph_recursively` so that the overall algorithm should be more readable and hopefully easier to understand, but at this point, I didn't find a good way to extract a method without having too much overload. This could be done in a later refactoring as an improvement.